// brightspace-auth-keys-dynamodb-store/src/public-key-table-setup.js
'use strict';

function createPublicKeyTable(db, tablename, readCapacity, writeCapacity) {
	var params = {
		AttributeDefinitions: [
			{
				AttributeName: 'PublicKey',
				AttributeType: 'S'
			}
		],
		KeySchema: [
			{
				AttributeName: 'PublicKey',
				KeyType: 'HASH'
			}
		],
		ProvisionedThroughput: {
			ReadCapacityUnits: readCapacity || 5,
			WriteCapacityUnits: writeCapacity || 5
		},
		TableName: tablename
	};

	return db.createTable(params).promise().then(() => {
		return setTtl(db, tablename);
	}).catch((e) => {
		// ResourceInUseException indicates the table already exists
		if (e.code !== 'ResourceInUseException') {
			throw e;
		}
	});
}

function setTtl(db, tablename) {
	var params = {
		TableName: tablename,
		TimeToLiveSpecification: {
			AttributeName: 'ExpiresAt',
			Enabled: true
		}
	};
	return db.updateTimeToLive(params).promise().catch(e => {
		// UnknownOperationException will be thrown for Dynamo Local
		if (e.code !== 'UnknownOperationException') {
			throw e;
		}
	});
}

module.exports = createPublicKeyTable;
