// brightspace-auth-keys-dynamodb-store/src/public-key-store.js
'use strict';

const AbstractPublicKeyStore = require('brightspace-auth-keys').AbstractPublicKeyStore;

const clock = require('./clock');

class DynamoPublicKeyStore extends AbstractPublicKeyStore {
	constructor(client, tableName) {
		super();
		this._client = client;
		this._tableName = tableName;
	}

	_storePublicKey(publicKey, expiresAt) {
		return this._client.putItem(
			{
				TableName: this._tableName,
				Item: {
					'PublicKey': {
						'S': publicKey
					},
					'ExpiresAt': {
						'N': '' + expiresAt
					}
				}
			}
		).promise();
	}

	_lookupPublicKeys() {
		const parameters = {
			TableName: this._tableName,
			FilterExpression: '#e > :expires_at',
			ExpressionAttributeNames: {
				'#e': 'ExpiresAt'
			},
			ExpressionAttributeValues: {
				':expires_at': {
					'N': '' + clock()
				}
			},
			ConsistentRead: true
		};

		return this._scanUntilComplete(null, parameters, []);
	}

	_scanUntilComplete(response, parameters, items) {
		let shouldContinue = false;

		// if we haven't had any results
		if (!response) {
			shouldContinue = true;
		}
		// if we had result on the last page
		else if (response.LastEvaluatedKey) {
			shouldContinue = true;
			parameters.ExclusiveStartKey = response.LastEvaluatedKey;
		}

		if (response && response.Items) {
			items = items.concat(response.Items.map(row => row.PublicKey.S));
		}

		if (shouldContinue) {
			return this._client.scan(parameters)
				.promise()
				.then(response => this._scanUntilComplete(response, parameters, items));
		} else {
			return Promise.resolve(items);
		}
	}
}

module.exports = DynamoPublicKeyStore;
module.exports.createTable = require('./public-key-table-setup');
