// brightspace-auth-keys/src/ec-key-generator.js
'use strict';

const { generateKeyPair, createPublicKey } = require('crypto');

const keyto = require('@trust/keyto');

const DEFAULT_CRV = 'P-256';

const CRV_TO_ALG = {
	'P-256': 'ES256',
	'P-384': 'ES384',
	'P-521': 'ES512'
};

const SUPPORTS_KEY_OBJECTS = typeof createPublicKey === 'function';
const PUBLIC_EXPORT_OPTIONS = {
	type: 'spki',
	format: 'pem'
};
const PRIVATE_EXPORT_OPTIONS = SUPPORTS_KEY_OBJECTS
	? null
	: {
		type: 'pkcs8',
		format: 'pem'
	};

function keygen(opts, kid) {
	return new Promise((resolve, reject) => {
		generateKeyPair('ec', {
			namedCurve: opts.crv,
			publicKeyEncoding: PUBLIC_EXPORT_OPTIONS,
			privateKeyEncoding: PRIVATE_EXPORT_OPTIONS
		}, (err, publicKey, privateKey) => {
			if (err) {
				return reject(err);
			}

			resolve([
				publicKey,
				privateKey
			]);
		});
	})
		.then(res => {
			return {
				jwk: Object.assign(keyto.from(res[0], 'pem').toJwk('public'), {
					kid,
					alg: CRV_TO_ALG[opts.crv],
					use: 'sig'
				}),
				signingKey: {
					kid,
					key: res[1],
					alg: CRV_TO_ALG[opts.crv]
				}
			};
		});
}

module.exports = keygen;
module.exports.normalize = function normalize(opts) {
	if (opts && typeof opts.crv !== 'undefined') {
		const crv = opts.crv;

		if (typeof crv !== 'string') {
			throw new TypeError(`"opts.ec.crv" should be a String. Got "${typeof crv}".`);
		}

		if (!CRV_TO_ALG[crv]) {
			throw new Error(`"opts.ec.crv" must be on of ${Object.keys(CRV_TO_ALG)}. Got "${crv}".`);
		}

		return { crv };
	}

	return { crv: DEFAULT_CRV };
};
